File: cpp/pineapple.cpp
#include "poker.cpp"

typedef vector<card> row;
typedef vector<row> board;

inline int sz(board &b) {
  return b[0].size() + b[1].size() + b[2].size();
}

board _board(row a, row b, row c) {
  return board({a, b, c});
}

void disp(row r, string sep=" ") {
  string ranks = "..23456789TJQKA", suits = "CDHS";
  cout << "ROW(";
  for (card c: r) {
    cout << ranks[c.Rank] << char(tolower(suits[c.Suit])) << " ";
  }
  cout << ")" << sep;
}

void disp(board B, string sep=" ", string sep2 = "\n") {
  cout << "BOARD(";
  for (row r: B) {
    cout << sep2;
    disp(r, "");
  }
  cout << ")" << sep;
}

bool isFL(hand h) {
  return h >= hand(PAIR, 12);
}

int royalty(int rid, hand h) {
  int T = h.Tier, V = int(h.Tiebreak);

  if (rid == 0) {
    if (T == PAIR) {
      return max(0, V - 5);
    } else if (T == THREE_OF_A_KIND) {
      return V + 8;
    }
  } else {
    if (rid == 1 && T == THREE_OF_A_KIND) {
      return 2;
    }

    if (T < STRAIGHT) return 0;
    int multiplier = rid == 1 ? 2 : 1;
    if (T == STRAIGHT) {
      return 2 * multiplier;
    } else if (T == FLUSH) {
      return 4 * multiplier;
    } else if (T == FULL_HOUSE) {
      return 6 * multiplier;
    } else if (T == FOUR_OF_A_KIND) {
      return 10 * multiplier;
    } else if (T == STRAIGHT_FLUSH) {
      return 15 * multiplier;
    } else if (T == ROYAL_FLUSH) {
      return 25 * multiplier;
    }
  }
  return 0;
}
int royalties(board B) {
  hand a = _hand(B[0]), b = _hand(B[1]), c = _hand(B[2]);
  if (a > b || b > c) {
    return 0;
  } else {
    int sum = royalty(0, a) + royalty(1, b) + royalty(2, c);
    return sum;
  }
}

File: src/share/pineapple.js
var p = require('./poker');
var U = require('./utils');
module.exports = (function() {
  if (!p.getPokerHand) {
    console.error("poker.js not included. pineapple.js will not run.");
    return;
  }

  var bonuses = new Map([
      [p.tiers.Straight, 2],
      [p.tiers.Flush, 4],
      [p.tiers.FullHouse, 6],
      [p.tiers.FourOfAKind, 10],
      [p.tiers.StraightFlush, 15],
      [p.tiers.RoyalFlush, 25],
    ])

  function topBonus(hand, wasFL) {
    var ret = 0;
    var value = Math.floor(hand.kickerValue); // this gives the main part of the hand (0-th kicker)
    var rankName = p.getPluralStr(p.numToRank(value));
    if (hand.tier == p.tiers.ThreeOfAKind) {
      return {
        name: 'Three of a Kind ' + rankName,
        royalty: value + 8,
        fl: true,
      }
    } else if (hand.tier == p.tiers.Pair && value >= 6) {
      return {
        name: rankName,
        royalty: Math.max(0, value - 5),
        fl: value >= 12 && !wasFL,
      }
    } else {
      return {
        name: hand.name,
        royalty: 0,
        fl: false,
      };
    }
  }

  function midBonus(hand, wasFL) {
    var royalty = 0;
    var fl = wasFL && (hand.tier >= p.tiers.FourOfAKind);
    if (bonuses.has(hand.tier)) {
      royalty = bonuses.get(hand.tier) * 2;
    } else if (hand.tier == p.tiers.ThreeOfAKind) {
      royalty = 2;
    }

    return { name: hand.name, royalty: royalty, fl: fl, }
  }

  function botBonus(hand, wasFL) {
    var royalty = 0;
    var fl = wasFL && (hand.tier >= p.tiers.FourOfAKind);
    if (bonuses.has(hand.tier)) {
      royalty = bonuses.get(hand.tier);
    }

    return { name: hand.name, royalty: royalty, fl: fl, }
  }

  function legal(top_, mid, bot) {
    return !p.betterPokerHand(mid, bot) && !p.betterPokerHand(top_, mid);
  }

  function scorePineapple(top_, mid, bot) {
    if (legal(top_, mid, bot)) {
      var a = topBonus(top_), b = midBonus(mid), c = botBonus(bot);
      return a.royalty + b.royalty + c.royalty;
    } else {
      return 0;
    }
  }

  var pp = {};
  pp.scorePineapple = scorePineapple;
  pp.topBonus = topBonus;
  pp.midBonus = midBonus;
  pp.botBonus = botBonus;
  pp.noBonus = {name: 'nobonus', royalty: 0, fl: false};
  pp.getBonus = function(hand, idx, wasFL) {
    return [topBonus, midBonus, botBonus][idx](hand, wasFL);
  }
  pp.fouled = function(board) {
    console.log('checking foul', board);
    var ph = p.getPokerHand;
    console.log('hands:', ph(board[0]), ph(board[1]), ph(board[2]));
    return !legal(ph(board[0]), ph(board[1]), ph(board[2]));
  };
  pp.matchup = function(here, there) {
    var ph = p.getPokerHand;
    var gt = p.betterPokerHand;
    return U.range(3).map(function(i) {
      var a = ph(here[i]), b = ph(there[i]);
      return gt(a, b) ? 1 : (gt(b, a) ? -1 : 0);
    });
  };

  function fillSuit(s) {
    if (s.length == 1) {
      s += 's';
    }
    return s;
  }

  pp.ss = function(s) {
    return p.getPokerHand(s.split(' ').map(fillSuit));
  };
  return pp;
})();
